### PR TITLE
Vue widgets

### LINKED ISSUE #7731: Drow down widget.

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Adds widgets:
- Checkbox (with sorting)
- Numeric slider
- Dropdown (accepting a list of strings)
  - Closes #7731
- Placeholder (underscore - has no actions)

### Important Notes
The widgets are currently added to every node, but are not synced with the yjs representation. This is intentional, as (afaict) the format for the AST representation is not yet finalized.

There are a number of design differences, for practical reasons:
- The dropdown now has a scrollbar.
  - As a side effect, the sort button needed to be moved left, to avoid overlapping with the scrollbar.
  - Note that it is *not* centered in the 8px horizontal padding. It is 4px wide, and has 1px left and 3px right padding: `.||||...`. (Note that the 8px horizontal padding from the design is retained.
- 4px of vertical padding has been inserted, so that there is *some* padding between the bubble for the selected item, and the outer dropdown container, when the first item is selected. Note that this is different to the 8px
- 16px of right margin has been inserted after every item. This is the same amount of padding that is added by the bubble. This means that the dropdown does not change in width when a long item is selected.

Design issues:
- The sort button for the dropdown overlaps the text
### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    </ul>
+  </div>
+</template>
+
+<style scoped>
+.Dropdown {
+  position: absolute;
+  user-select: none;
+  overflow: auto;
+  clip-path: inset(0 round 8px);
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: use `border-radius` instead. Using clip-paths inside scaled divs tends to make text rendering blurry.

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  height: 136px;
+  -ms-overflow-style: none;
+  scrollbar-width: none;
+}
+
+.Dropdown::-webkit-scrollbar {
+  display: none;
+}
+
+.list {
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: This `ul` doesn't have the list markers styled, making it visible by default. Needs `list-style-type: none;`.

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  position: relative;
+  color: var(--color-text-light);
+  padding-left: 8px;
+  padding-right: 8px;
+  width: min-content;
+}
+
+.list > * {
+  position: relative;
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: Why is this necessary?

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      <template v-for="[value, index] in sortedValuesAndIndices" :key="value">
+        <li v-if="value === selectedValue">
+          <div class="selected-item"><span v-text="value"></span></div>
+        </li>
+        <li v-else class="selectable-item button" @click="emit('click', index)">
+          <span v-text="value"></span>
+        </li>
+      </template>
+      <img
+        class="sort button"
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: This `sort` icon looks very bad when it overlaps with text. Since it's not clear how to fix it, please raise this as requiring clarification on the design. Also currently it is positioned within the scrolled container, which is also likely not the desired behavior. I imagine something like `position: fixed` is needed.

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+.Dropdown {
+  position: absolute;
+  user-select: none;
+  overflow: auto;
+  clip-path: inset(0 round 8px);
+  width: min-content;
+  top: 100%;
+  margin-top: 4px;
+  height: 136px;
+  -ms-overflow-style: none;
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: Unnecessary. We don't care about legacy IE/edge.

File: app/gui2/src/components/widgets/Dropdown.vue [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  scrollbar-width: none;
+}
+
+.Dropdown::-webkit-scrollbar {
+  display: none;
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Dropdown.vue]: We need a scrollbar. We can use native ones, but it should be styled to mimic the one on the design, at least as close as CSS will allow. We might eventually move on to custom scrollbar components, but we have to have something functional until that happens.

File: app/gui2/src/components/widgets/Slider.vue [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    const newValue = props.min + Math.round(fraction * (props.max - props.min))
+    emit('update:modelValue', newValue)
+  }
+}
+
+useDocumentEvent('mouseup', () => {
+  isDragging.value = false
+})
+
+useDocumentEventConditional('mousemove', isDragging, (event) => {
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Slider.vue]: We already have `usePointer` for handling pointer drag interactions.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 3 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: i usually add it because absolute/relative both create a stacking context, so when something has a position absolute sibling that overlaps it, then the text may disappear under the sibling. in this case it's probably not needed though,. I'll try removing it and report back

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 2 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: the list markers don't seem to show up by default, for some reason - but definitely a good idea to be extra sure.

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 6 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: i don't think there are any scrollbars anywhere in the design actually - but can definitely add one

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 4 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: makes sense, i'm not sure where's the most appropriate place to raise it though

**Comment by somebody1234** [comment]
@Frizi all issues *should* be addressed - except the one about the sort button, since that requires design clarifications

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 4 above)
Comment (Frizi) [reply app/gui2/src/components/widgets/Dropdown.vue]: You can do two things in such cases, best both.
1. Ask @wdanilo for long term design clarification and work out something to implement right now.
2. Try to implement something sensible if you have any quick ideas to try. In this case, putting even a small background-color matched shadow below the icon would make it more legible when it overlaps the text. Something along those lines would probably do for the time being and allow us to move forward.

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 6 above)
Comment (Frizi) [reply app/gui2/src/components/widgets/Dropdown.vue]: When in doubt, look at what's implemented in current IDE. We have scrollbars. The design simply missed them/didn't show them in shown state. This could also be raised with @wdanilo to make sure it's visible in figma in at least one place.

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 3 above)
Comment (Frizi) [reply app/gui2/src/components/widgets/Dropdown.vue]: Stacking contexts or other such constructs should be added as required, not "just in case". Try to keep the styles to minimum that is necessary at present.

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 2 above)
Comment (Frizi) [reply app/gui2/src/components/widgets/Dropdown.vue]: They do show up. In chrome, the dots are present but fully clipped out by the scroll container. But in firefox, you can actually still see them.
image

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 3 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: yeah i think i added it out of habit? note that this *is* needed for the toolbar and circular menu for visualizations, since those *do* have a 64px background blur.

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 2 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: ah, good to know

File: app/gui2/src/components/widgets/Dropdown.vue (hunk 6 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Dropdown.vue]: i don't get scrollbars in ide unfortunately :(
unless it's a fairly recent feature? or unless we're supposed to use the same scrollbars as in the CB

image

Comment (Frizi) [reply app/gui2/src/components/widgets/Dropdown.vue]: I guess the dropdowns are allowed to be quite long before they show the scrollbar. But it is the same component as in the CB indeed.

**Comment by Frizi** [comment]
This PR has a bunch of repeated things with #7726, so let's merge the other one first, and then I'll take another look here.

File: app/gui2/src/components/widgets/Placeholder.vue [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<script setup lang="ts">
+import UnderscoreIcon from '@/assets/icons/underscore.svg'
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Placeholder.vue]: This is no longer needed, we have `SvgIcon` component. Just move the added icons to the common icon file.

File: app/gui2/src/components/widgets/Slider.vue [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+import { ref, computed } from 'vue'
+import { PointerButtonMask, usePointer } from '@/util/events'
+
+const props = defineProps<{ modelValue: number; min: number; max: number }>()
+const emit = defineEmits<{ 'update:modelValue': [modelValue: number] }>()
+
+const sliderNode = ref<HTMLElement>()
+
+/** The flag in `event.buttons` representing the left mouse button. */
+const BUTTON_LEFT_FLAG = 1
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Slider.vue]: Unused const

File: app/gui2/src/components/widgets/Slider.vue [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+const dragPointer = usePointer((pos, _, eventType) => {
+  if (sliderNode.value == null) {
+    return
+  }
+  if (eventType !== 'move') {
+    return
+  }
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Slider.vue]: Why is this necessary? Clicling on the slider should probably update the value just as well.

File: app/gui2/src/components/widgets/Slider.vue [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    return props.modelValue
+  },
+  set(value) {
+    emit('update:modelValue', value)
+  },
+})
+</script>
+
+<template>
+  <div ref="sliderNode" class="Slider" @="dragPointer.events">
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Slider.vue]: I like this `@=""` syntax, but have you managed to make it work with eslint somehow? Last time I checked this caused errors. If you have a way to get rid of them in this case while keeping property validation enabled, i'm fine with using it.

File: app/gui2/src/components/widgets/Slider.vue [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  position: relative;
+  user-select: none;
+  display: flex;
+  justify-content: space-around;
+  background: var(--color-widget);
+  border-radius: var(--radius-full);
+  width: 56px;
+}
+
+.Slider > .fraction {
```
Comment (Frizi) [inline app/gui2/src/components/widgets/Slider.vue]: Nested selector not necessary. Just using `.fraction` is enough thanks to scoping.

**Comment by sylwiabr** [comment]
@Frizi the mentioned PR was merged, anything else blocking a review?

**Review by Frizi — APPROVED** [review]
Small nits/questions remaining, but looks good. Also make sure that it passes the linter.

File: app/gui2/src/components/widgets/Slider.vue (hunk 11 above)
Comment (somebody1234) [reply app/gui2/src/components/widgets/Slider.vue]: oh nice catch, thanks
